app/api-client.js

```diff
@@ -1,170 +1,291 @@
+  }
 
-    return Object.assign(defaults, options);
+  delete(endpoint) {
+    return this.request(endpoint, 'DELETE');
   }
 
-  function endpoint_url(endpoint) {
+  patch(endpoint, body = '') {
+    return this.request(endpoint, 'PATCH', body);
+  }
+
+  put(endpoint, body = '') {
+    return this.request(endpoint, 'PUT', body);
+  }
+
+  endpoint_url(endpoint) {
     if (!endpoint) {
       throw new Error('Missing endpoint');
     }
 
+    let suffix = '';
     if (!endpoint.endsWith('/')) {
-      endpoint += '/';
+      suffix = '/';
     }
 
-    return url.resolve(config.api.base_url, endpoint);
+    return url.resolve(this.base_url, endpoint + suffix);
   }
+}
+
+
+const api = new APIClient(config.api);
+api.auth = new JWTAuth();
+
+exports.api = api;
+
 
-  function basic_auth() {
-    return 'Basic ' + new Buffer(
-      config.api.username + ':' + config.api.password).toString('base64');
+const model_proxy = {
+  get: (model, property) => {
+    if (Reflect.has(model, property)) {
+      return Reflect.get(model, property);
+    }
+    if (model.data && property in model.data) {
+      return model.data[property];
+    }
+    return undefined;
+  },
+};
+
+
+class ModelSet extends Array {
+  constructor(Model, data = []) {
+    super(...data.map(obj => new Model(obj)));
+    this.model = Model;
   }
 
-  function jwt_auth() {
-    return `JWT ${token}`;
+  exclude(other) {
+    let others = other;
+    if (other instanceof this.model) {
+      others = [other];
+    }
+
+    const pks = others.map(instance => instance[this.model.pk]);
+
+    return this.filter(instance => !pks.includes(instance[this.model.pk]));
   }
+}
 
-  let api = {};
+exports.ModelSet = ModelSet;
 
-  api.set_token = function (jwt) {
-    token = jwt;
-  };
+class Model {
+  constructor(data) {
+    this.data = data;
+    return new Proxy(this, model_proxy);
+  }
 
-  api.unset_token = function () {
-    token = 'invalid token';
-  };
+  static get pk() {
+    return 'id';
+  }
 
-  api.list_users = function () {
-    return api_request({ endpoint: 'users' });
-  };
+  static list() {
+    return api.get(this.endpoint)
+      .then(result => new ModelSet(this.prototype.constructor, result.results));
+  }
 
-  api.get_user = function (id) {
-    return api_request({ endpoint: `users/${id}` });
-  };
+  static get(id) {
+    return api.get(`${this.endpoint}/${id}`)
+      .then(data => new this.prototype.constructor(data));
+  }
 
-  api.add_user = function (user) {
-    return api_request({ method: 'POST', endpoint: 'users', resource: user });
-  };
+  static delete(id) {
+    return api.delete(`${this.endpoint}/${id}`);
+  }
 
-  api.users = {
-    list: api.list_users,
-    get: api.get_user,
-    add: api.add_user,
-  };
+  create() {
+    return api.post(this.constructor.endpoint, this.data)
+      .then(data => new this.constructor(data));
+  }
 
-  api.list_apps = function () {
-    return api_request({ endpoint: 'apps' });
-  };
+  replace() {
```

Comment:
Is this `replace()` method used anywhere? Also, what is it doing? Is it just to make a `PUT` request to the API? We may not need it (yet?).

Reply:
Yeah, it's not used anywhere yet.